src/main/java/net/starschema/clouddb/cmdlineverification/Oauth2Bigquery.java

```diff
@@ -219,29 +220,41 @@ public static Bigquery authorizeviaservice(String serviceaccountemail,
                                                Integer connectTimeout,
                                                Integer readTimeout) throws GeneralSecurityException, IOException {
         logger.debug("Authorizing with service account.");
-
         List<String> scopes = new ArrayList<String>();
         scopes.add(BigqueryScopes.BIGQUERY);
         // don't have access to DriveScopes without requiring the entire google drive sdk.
         scopes.add(DRIVE_SCOPE);
-        GoogleCredential.Builder builder = new GoogleCredential.Builder()
-                .setTransport(CmdlineUtils.getHttpTransport())
-                .setJsonFactory(CmdlineUtils.getJsonFactory())
-                .setServiceAccountId(serviceaccountemail)
-                        // e-mail ADDRESS!!!!
-                .setServiceAccountScopes(scopes);
-                        // Currently we only want to access bigquery, but it's possible
-                        // to name more than one service too
-
-        if (password == null) {
-            builder = builder.setServiceAccountPrivateKeyFromP12File(new File(keypath));
+
+        GoogleCredential credential = null;
+
+        // If .p12 use builder to create credential
+        if (Pattern.matches(".*\\.p12$", keypath)) {
+            GoogleCredential.Builder builder = new GoogleCredential.Builder()
+                    .setTransport(CmdlineUtils.getHttpTransport())
+                    .setJsonFactory(CmdlineUtils.getJsonFactory())
+                    .setServiceAccountId(serviceaccountemail)
+                    // e-mail ADDRESS!!!!
+                    .setServiceAccountScopes(scopes);
+            // Currently we only want to access bigquery, but it's possible
+            // to name more than one service too
+
+            if (password == null) {
+                builder = builder.setServiceAccountPrivateKeyFromP12File(new File(keypath));
+            }
+            else {
+                PrivateKey pk = getPrivateKeyFromCredentials(keypath, password);
+                builder = builder.setServiceAccountPrivateKey(pk);
+            }
+            credential = builder.build();
         }
+        // If not .p12 must be .json
         else {
-            PrivateKey pk = getPrivateKeyFromCredentials(keypath, password);
-            builder = builder.setServiceAccountPrivateKey(pk);
+            // For .json load the key via credential.fromStream
+            File jsonKey = new File(keypath);
+            InputStream inputStream = new FileInputStream(jsonKey);
+            credential = credential.fromStream(inputStream, CmdlineUtils.getHttpTransport(), CmdlineUtils.getJsonFactory()).createScoped(scopes);
```

Comment:
What happens here if a malformed JSON file is passed? Does this fail gracefully?

Reply:
Does the JSON file include the service account email? Do we need to worry about setting one? Is the service account email parameter just ignored here?

Reply:
```
{
  "type": "service_account",
  "project_id": "... PROJECT ID",
  "private_key_id": "... PK ID",
  "private_key": "... PRIVATE KEY",
  "client_email": "... CLIENT EMAIL",
  "client_id": "... CLIENT ID",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://accounts.google.com/o/oauth2/token",
  "auth_provider_x509_cert_url": "... CERT GOES HERE",
  "client_x509_cert_url": "... CERT GOES HERE"
}
``` 
format of a json key with all sensitive info removed.

When you generate they key it has to be attached to an email.

Reply:
```
projectid=disco-parsec-659
type=service
user= EMAIL FOR SERVICE ACCOUNT 
password= PATH TO THE JSON KEY ( same format as above )
dataset= DATASET 
transformquery=true
```

This is the `.properties` file I am testing with for json

Reply:
Without an email - it will fail with a query  when you run the query but still create a connection. `jdbc.BQSQLException` 
 In helltool looks like this - 

![screen shot 2017-10-06 at 12 09 06 pm](https://user-images.githubusercontent.com/5461381/31294314-3cae4804-aa8f-11e7-8be0-e4c12f7acbae.png)

With a completely borked json keyfile it will throw an IO exception. Which looks like this in helltool.

![screen shot 2017-10-06 at 12 10 19 pm](https://user-images.githubusercontent.com/5461381/31294336-5c24d8ec-aa8f-11e7-8c02-d3b06b91064e.png)